src/pudl/harvest/harvest.py

```diff
@@ -0,0 +1,1054 @@
+            index and values, respectively,
+            of the :class:`pd.Series` returned for each column.
+            By default, the first value is `x.name`
+            (the values of columns `by` for that row group),
+            and the second is the original error.
+
+    Returns:
+        Aggregated dataframe with `by` columns set as the index and
+        (if `errors`='report') an error report with a :class:`pd.Series`
+        for each column where errors occured.
+
+    Examples:
+        >>> import pandas as pd
+        >>> df = pd.DataFrame({'x': [0, 0, 1, 1], 'y': pd.Series([2, 2, 2, 3], dtype='Int64')})
+        >>> df.index = [0, 0, 0, 1]
+        >>> base = dict(df=df, by='x', aggfuncs={'y': unique})
+        >>> groupby_apply(**base, errors='raise')
+        Traceback (most recent call last):
+          ...
+        ValueError: Could not aggregate y at x = 1: Not unique.
+        >>> groupby_apply(**base, errors='coerce')
+              y
+        x ...
+        0     2
+        1  <NA>
+        >>> _, report = groupby_apply(**base, errors='report')
+        >>> report['y']
+        x
+        1    Not unique.
+        dtype: object
+        >>> errorfunc = lambda x, e: (x.name, as_dict(x))
+        >>> _, report = groupby_apply(**base, errors='report', errorfunc=errorfunc)
+        >>> report['y']
+        x
+        1    {0: [2], 1: [3]}
+        dtype: object
+    """
+    groupby = df.groupby(by)
+    data_columns = [col for col in df.columns if col not in by]
+    series = {}
+    reports = {}
+    for col in data_columns:
+        report = []
+        aggfunc = aggfuncs[col]
+        if errors == "raise":
+            msg = f"Could not aggregate {col} at {by} = {{x.name}}: {{e}}"
+            wrapper = try_aggfunc(aggfunc, method="raise", error=msg)
+        elif errors == "coerce":
+            wrapper = try_aggfunc(aggfunc, method="coerce")
+        elif errors == "report":
+            if errorfunc is None:
+
+                def errorfunc(x, e):
+                    return x.name, str(e)
+
+            # NOTE: Appending to list in scope ~10% faster than passing list as argument
+            # aggfunc = try_aggfunc(aggfunc, method='append', error=errorfunc)
+            def wrapper(x):
+                try:
+                    return aggfunc(x)
+                except ValueError as e:
+                    report.append(errorfunc(x, e))
+                    return np.nan
+
+        ds = groupby[col].apply(wrapper)
+        if str(ds.dtype) != str(df[col].dtype):
+            # Undo type changes triggered by nulls
+            ds = ds.astype(df[col].dtype)
+        if report:
+            report = pd.Series(dict(report))
+            keys = ds.index.names
+            if report.index.nlevels == len(keys):
+                report.index.names = keys
+            reports[col] = report
+        series[col] = ds
+    if errors == "report":
+        return pd.DataFrame(series), reports
+    return pd.DataFrame(series)
```

Comment:
can we just return both at all times even if `reports= None` or `[]`

Reply:
Yeah, that was a bad API decision on my part. `groupby_apply()` and `groupby_aggregate()` now always have the same return data type (`pd.DataFrame`, `dict`), with `reports={}` by default.